### ISSUE TITLE
Advanced MERGE incremental settings

### ISSUE THREAD

**Issue by mi-volodin** [databricks/dbt-databricks#245]
### Describe the feature
Currently merge **conditions** are not supported in merge strategy.

If following the Databricks terminology, merge conditions are `matched_conditions` and `not_matched_conditions` in `MERGE` definition.

Taken from DBricks docs:
```sql
MERGE INTO target_table_name [target_alias]
   USING source_table_reference [source_alias]
   ON merge_condition
   { WHEN MATCHED [ AND matched_condition ] THEN matched_action |
     WHEN NOT MATCHED [ AND not_matched_condition ] THEN not_matched_action } [...]
```

As for now the full power of `MERGE` statement is not supported.

### Describe alternatives you've considered
Writing customized adapter with own `merge` macro to perform such operations. Add conditions to configuration for model.

### Additional context
Databricks is more and more often used as a tool to support a variety of data operations that was previously relied on some analytical MPPs. Specifically preparing some DWH-like normalized models for long-term entity-oriented storage and creating non-denormalized data marts.

One of the drawbacks of the merge is that update operation is basically DELETE + INSERT, so it creates some additional expenses for each record that you are updating. So it is widely used techniques to minimise such updates by additional conditions which may apply only for already observed records, and thus can't be defined as `merge predicates` (in dbt terminology).

Moreover, there are business cases that require that. For example, I am tracking some object and its properties as an entry of the form `<o, P>`. If I observe some specific instance of such record (`<1, A>`) I may want to skip all consequent updates if the properties are not changing. And do the update only if I see `<1, B>` (for example).

### Who will this benefit?
Anyone who uses MERGE as a strategy for model updates and requires additional business checks for the records before making updates / inserts.

### Are you interested in contributing this feature?
Probably, but due to variety of possible solutions I will appreciate the solution design suggestion to be discussed with maintainers.

**Comment by adrianabreu** [comment]
Any updates on this? Seems an interesting feature

**Comment by mi-volodin** [comment]
No updates. Just to note, that there's a work-around that currently might work - with `dbt` you have to manually rebuild the whole update set by your model. That makes the model code unbearably huge, but it works.

For me this just outlines the cases when `dbt` modelling approach is not nice. Also the usage of null-aware joins I think is missing as a feature, as `MERGE` is frequently used for denormalized "cubes" where missing dimensions are allowed to be `NULL`. So using `<=>` on merge predicates instead of `=` is also required. And I don't know a workaround for this.

**Comment by mi-volodin** [comment]
Okay, now I feel the urge to make a try and come up with my own solution.

What I generally propose as the capabilities:

- Add `match_conditions` and `non_match_conditions` optional parameters that will hold explicit string for predicate. Probably I will need help to understand how to protect string injection here.
- I am also thinking on how I can better control the merge capabilities. Today we have obligatory insert + update. ALso no support for "not matched by source". I want to add this capabilities as well.

I see that there's a connected context in #645 . @IvanDerdicDer are you really on it? Do you see any potential conflicts with what is proposed above?

Also, @benc-db WDYT about creating the umbrella issue that will highlight the gap between Databricks merge capabilities and dbt implementation, so we can track the progress.

**Comment by IvanDerdicDer** [comment]
Hey, 

Still not working on it, collage got the priority.  I was thinking mostly the same but just for the match statement, rest of it sounds good. 

I did a proof of concept at work when we were developing delta data delivery method, and we tried to create a custom merge connector. That didn't work because there is no support for custom connectors.
In short what I did was that I just copied the merge_update_columns function and modified it to return an update condition ie. one column not equal. Haven't tested it because custom incremental strategies don't work but the generated SQL looked good.

I believe the same approach can be utilized for everything you described. Also I believe that you don't have to worry about string injection, none of the functions that take merge_update_columns, or merge_exclude_columns  take that into account. As well most of dbt code is generally written by hand and not auto generated by an API or such.

**Comment by mi-volodin** [comment]
Ok, I also will wait for @benc-db response.

I also see a potential issue with the matched conditions, as they have to be explicit. And therefore should use aliases.

And here it seems that it creates [unnecessary bound ](https://github.com/databricks/dbt-databricks/blob/3cbf7dc45c085ebf3c6c73c403541110f08d1e4d/dbt/include/databricks/macros/materializations/incremental/strategies.sql#L100-L104) with the aliases in the code.

So if I understand correctly now one should write

`DBT_INTERNAL_DEST.A > DBT_INTERNAL_SOURCE.A`

I know that it is possible to workaround this with particular placeholders that we can automatically transcribe. For instance, allow to write `dest.A > source.A`... But I am not sure it is solid.

**Comment by benc-db** [comment]
@mi-volodin at the moment I don't have capacity myself, but I'm happy to help you develop a solution.  I would recommend after your initial investigation that you write your proposed approach here (what files you plan on modifying, and how), and I can give you some initial feedback before the PR.  Once you put a PR out, I can pull/push to get our integration tests to run on it.

**Comment by IvanDerdicDer** [comment]
I don't see a need for transcribing DBT_INTERNAL_SOURCE/DEST seem fine to me. Plus they would be autogenerated, at least that is what I had in mind. You would pass a list of columns that need to be different so that update occurs. To achieve that i modified
[get_merge_update_set](https://github.com/databricks/dbt-databricks/blob/3cbf7dc45c085ebf3c6c73c403541110f08d1e4d/dbt/include/databricks/macros/materializations/incremental/strategies.sql#L107C1-L119C15) function. Replaced '=' with '<>', and ',' with 'AND'. That works for wast majority of upserts/merge operations I believe.

From what I gather you want to let the user write a SQL string as a condition. I don't think that is necessary, a list of columns should suffice. But if you decide to go the SQL string route then the same as for the incremental predicate the user should write DBT_INTERNAL_SOURCE/DEST.

**Comment by mi-volodin** [comment]
@IvanDerdicDer the `matched conditions` could be an arbitrary predicate, very complex in general case. So I probably don't follow, but I see no way to autogenerate it.

For example, if I want something like this

```sql
when matched 
    and SRC.change_ts > TGT.change_ts
    and hash(SRC.a, SRC.b) <> hash(TGT.a, TGT.b)
```

I will have to either use `DBT_INTERNAL_SOURCE` here as an alias (which I don't think is smart), or introduce special hardcoded aliases like I used above which will be then mapped to `DBT_INTERNAL_SOURCE/TARGET`.

**Yes**, as an alternative - it is possible to limit the space of predicate that could be used, and then use generators.

@benc-db got it. Will get back.

**Comment by mi-volodin** [comment]
@benc-db I am starting with a higher level vision, will be glad to specify more after the discussion.

# Features logical descriptions

In general I think it make sense to extend the functionality with the following features

- [ ] Allow definition of `matched_conditions` as an explicit predicate
- [ ] Allow definition of `non_matched_conditions` as an explicit predicate
- [ ] Allow to skip `WHEN MATCHED` by adding `skip_matched: true`
- [ ] Allow to skip `WHEN NOT MATCHED` by adding `skip_non_matched: true`

Ideally it would be great to also extend for support of `WHEN NOT MATCHED BY SOURCE` condition, but let's keep it out of scope due to complexity. Though, I would keep it in mind, to narrow the solution space for better compatibility.

## The `matched_conditions` and `non_matched_conditions`

Both of the features will expect the following:

- The new keys are introduced in the config, both `string` like.
- The changes to (at least) [strategies.sql](https://github.com/databricks/dbt-databricks/blob/3cbf7dc45c085ebf3c6c73c403541110f08d1e4d/dbt/include/databricks/macros/materializations/incremental/strategies.sql), specifically to `databricks__get_merge_sql` macro.
- Changes to `DatabricksConfig`, I assume it's in [impl.py](https://github.com/databricks/dbt-databricks/blob/3c4f972ed1b0c781b9fcc4be84f05d335b85bd69/dbt/adapters/databricks/impl.py#L94-L105).

Important note is that for now I don't see any other way but to allow explicit predicates, where user should use source and target aliases to define the logic. The surfaced solution is that we allow to use specific aliases like `s. / t.` or `src./tgt.` (maybe several) to describe logic and then transcode it to this "DBT_INTERNAL_SOURCE" thingy after. 

If we keep this idea - that will require 1 extra macro in `strategies.sql`.

Also I don't know if `DBT_INTERNAL_*` prefix is required for something specific, otherwise we could just replace it with `SRC/TGT` and allow the explicit usage of it.

## The `skip_` flags

First of all, I understand that it is not, probably, very elegant. As ideally we actually should expect the user to ASK for a certain step. And why we perform both insert and update by default is probably the `dbt` standard (but not the SQL style).

So if we later go for `WHEN NOT MATCHED ON SOURCE` - the implementation will have to differ with insert/update blocks.

Nevertheless, I would be glad to have that logic implemented, so we can avoid ugly `dbt` hacks with `left anti join` "if target exists" to avoid updates.

From the implementation perspective I also see for now that it can be done on the `databricks__get_merge_sql` macro level and will require additional keys in `config`.

## Summary of changes

- New keys in the `DatabricksConfig` in `impl.py` + corresponding validations if needed:
  - `matched_conditions: Optional[str]`
  - `non_matched_conditions: Optional[str]`
  - `skip_matched_step: Optional[bool] = False`
  - `skip_non_matched_step: Optional[bool] = False`
- Macro extension `databricks__get_merge_sql` in `strategies.sql` that utilises new keys and introduces the logic.
- **Documentation.** Obviously new functionality have to be included there.
- **Tests.** To cover the introduced use cases.

## Questions to clarify

- As unexperienced `dbt` internals / adapters developer I don't see any other points where I have to introduce changes. Would be glad if someone can highlight some missed complexity.
- I am not sure if we need to care about security... The example with `matched_conditions` might allow string injections. There are cases when it is dangerous, but they are anyway superseded by the ability to explicitly put any code in the model. Thus - I ignore that risk.
- There's an `options` key in the config. It is not currently used in `merge` and I am not sure if I need to move certain config keys in the `options` dict actually.

**Comment by benc-db** [comment]
>Also I don't know if DBT_INTERNAL_* prefix is required for something specifi

To the best of my knowledge, these are just aliases assigned to help the readability of the generated SQL.  You could use a different placeholder, it will just need to be documented.

>I am not sure if we need to care about security... The example with matched_conditions might allow string injections. There are cases when it is dangerous, but they are anyway superseded by the ability to explicitly put any code in the model. Thus - I ignore that risk.

Agreed, safe to ignore as we already allow users to write arbitrary SQL in their dbt project.

>There's an options key in the config. It is not currently used in merge and I am not sure if I need to move certain config keys in the options dict actually.

I believe this is for OPTIONS clause on table create, which are nearly the same as TBLPROPERTIES, but I think older, Hive-related; should not be reused.

Any new config that is not defined by dbt-core can be read from model.config.extra dict, so you don't strictly need to modify DatabricksConfig either, but it's probably a good idea for visibility sake.  I believe adding them to DatabricksConfig makes them available in config, rather than config.extra.

Could you give a couple of examples of config and what the translated SQL should like?

**Comment by mi-volodin** [comment]
> Any new config that is not defined by dbt-core can be read from model.config.extra dict, so you don't strictly need to modify DatabricksConfig either, but it's probably a good idea for visibility sake. I believe adding them to DatabricksConfig makes them available in config, rather than config.extra.

Then it probably makes sense to prototype first with `config.extra` and if we will not receive any objections from the review - extend the `config`. For users it will make no difference if I understand correctly.

> Could you give a couple of examples of config and what the translated SQL should like?

Yeah, let me prepare few examples.

**Comment by mi-volodin** [comment]
@benc-db 
I generated couple of examples, will attach the file to save space, but here are some of them.

Example with step skip.
```sql
/*
    skip_matched_step: True,
    skip_not_matched_step: False,
    matched_conditions: 'None',  
    not_matched_conditions: 'None',
*/

merge into 
    TABLE1 as tgt
using 
    TABLE2 as src
on 
    predicates ...
when not matched
then insert >>get_merge_insert(on_schema_change, source_columns) <<
```

Example with matched conditions
```sql
/*
    skip_matched_step: False,
    skip_not_matched_step: False,
    matched_conditions: 'src.updated_at_ts > tgt.updated_at_ts or tgt.updated_at_ts is NULL',  
    not_matched_conditions: 'some_attribute > 0',
*/

merge into 
    TABLE1 as tgt
using 
    TABLE2 as src
on 
    predicates ...
when matched
    and src.updated_at_ts > tgt.updated_at_ts or tgt.updated_at_ts is NULL
then update set 
     >> get_merge_update_set(update_columns, on_schema_change, source_columns)<<
when not matched
    and some_attribute > 0
then insert >>get_merge_insert(on_schema_change, source_columns) <<
```

This is how it works right now
```sql
/*
    skip_matched_step: False,
    skip_not_matched_step: False,
    matched_conditions: 'None',  
    not_matched_conditions: 'None',
*/

merge into 
    TABLE1 as tgt
using 
    TABLE2 as src
on 
    predicates ...
when matched
then update set 
     >> get_merge_update_set(update_columns, on_schema_change, source_columns)<<
when not matched
then insert >>get_merge_insert(on_schema_change, source_columns) <<
```

[output.txt](https://github.com/user-attachments/files/15930021/output.txt)

**Comment by mi-volodin** [comment]
One of the problem I don't know how to elegantly fight with - is that these aliases. In my example I used `src` and `tgt` to distinguish between the source and destination, but what I initially thought I would do was something like this

```python
{{ conditions_str | regexp_replace('(^| )s[\w]+\.', 'DBT_INTERNAL_SOURCE.', ...flags...) }}
```
i.e. we allow to write any alias like `source., s., Src.` etc and replace it.

But appears that for this I will have to register a custom filter for Jinja, which I would prefer to avoid to keep complexity lower.

**Comment by benc-db** [comment]
> But appears that for this I will have to register a custom filter for Jinja, which I would prefer to avoid to keep complexity lower.

Anything that is not just formatting I would put in the Adapter class and write it/unit test it in python.  I hate Jinja personally :P.  I wish we had better mechanisms for exposing logic than just always throw in the adapter class, but using Python will make testing/maintenance simpler in the long run.  Also, if you look at `tests/unit/macros`, I've tried to make it more painless to tests Jinja macros, but it's still a bigger lift than doing the same in python.

**Comment by benc-db** [comment]
I think you're in a good place to do a PR.  I think I will understand the nuances better once I see some tests :).

**Comment by mi-volodin** [comment]
> Anything that is not just formatting I would put in the Adapter class and write it/unit test it in python.

Gosh... it's my Friday evening dumbness. Why on earth haven't I thought about Python side implementation 🤦🏻‍♂️ I also hate Jinja.

Ok, I will need some time to implement a draft and go through all contributions preparation steps. And of course pass through test 😄

**Comment by benc-db** [comment]
@mi-volodin take your time.  This seems like a great include for 1.9, which I'm thinking will land sometime in July.  I appreciate your effort/contribution!

**Comment by mi-volodin** [comment]
@benc-db I ran functional tests on all-purpose cluster (the only one option I have now).
And 6 tests in total are failing. 
<img width="613" alt="image" src="https://github.com/databricks/dbt-databricks/assets/8374532/f703e353-e8c5-46cd-a3c3-c980eefd3e85">

No code change done yet. Since it is likely not OK, I wanted to ask for a quick tip from an experienced maintainer. Probably I missed some requirement for DBR Runtime, for example.

I will dig into it anyway, but hopefully you can help me to save some time.

**Comment by benc-db** [comment]
@mi-volodin if you want to email me your cluster config, and we can compare notes?  ben.cassell@databricks.com.  Also, good call running all of the integration tests before code change, cause otherwise we'd be running down dead ends :)

**Comment by kmarq** [comment]
@mi-volodin will you be adding support for the delete option of the merge statement as well? I know that's something we were just looking to do but it looks like you are well ahead of us with your ongoing changes.

**Comment by mi-volodin** [comment]
@kmarq Yes, I was thinking about this, as the next step.
Also there's a request for SCHEMA EVOLUTION in #707, which also looks relatively straightforward.

But I would certainly commit only once I can achieve the first changes passing the tests.

@benc-db sent you a mail.

**Comment by mi-volodin** [comment]
Short update. Tests are still failing, but I have studied the logs and found the following reasons:

1. One test is failing, because it tries to store the data in s3 bucket (which is non-existent). I think it is fine, because it is not supposed to work in my case. Probably that test should not be included in my test-set (General purpose cluster).
2. Other 5 tests are failing because of the same error of duplicated reference to `dbt_unique_key`. My hunch is that it is because of that peace of code that snapshot macro generates:

```sql
    snapshotted_data as (

        select *,
            id as dbt_unique_key

        from `hive_metastore`.`snapshots`.`target_snap`
        where dbt_valid_to is null

    ),
```
And the reason for failure is that there is already a `dbt_unique_id` in the `target_snap`. Thus in `snapshotted_data` CTE there are exactly 2 `dbt_unique_id` attributes.

Databricks allows this (at least in Runtimes < 15)
<img width="416" alt="image" src="https://github.com/databricks/dbt-databricks/assets/8374532/f9615df2-12a7-437d-9c40-efd700306bbc">

@benc-db I need an advice if I need to deal with this. Seems also an interesting problem, but it also might be related to some of my local policies, so if the test passes during pull request review - I can try to just ingore that precise reason in the context of my local dev tests.

**Comment by benc-db** [comment]
> One test is failing, because it tries to store the data in s3 bucket (which is non-existent). I think it is fine, because it is not supposed to work in my case. Probably that test should not be included in my test-set (General purpose cluster).

This is almost certainly fine.  External location tests should happen with all types of compute, I believe, but they require setup outside of Databricks.

**Comment by kmarq** [comment]
@mi-volodin I was wondering if you have any thoughts on this being near completion? Is there anything that I could help assist with testing?

**Comment by mi-volodin** [comment]
@kmarq Sorry, actually I had not much time working on this. And since it is my first PR in this repo, getting it to the point even with minor changes is tricky. 

Let's collaborate! 
I think we can focus first on adding flags and cover it with tests. Then I finally become acquainted with the whole process, I will add the more challenging change with match conditions.

WDYT? 

p.s. @benc-db haven't found it in the rules. Is it okay to open WIP Draft PR here, or is it discouraged?

**Comment by benc-db** [comment]
@mi-volodin absolutely open a draft :)

**Comment by sabrinas-tuidi** [comment]
Hi @mi-volodin, 
We are currently running our dbt project on pre-released version 1.9.0, which includes your contribution for matched and not matched conditions. This is super useful so thank you so much for working on this! 
We are currently facing a blocker since, due to the nature of our workflow, we cannot perform a direct delete action when not matched by source. Instead, we need to run an update set on a flag that indicated the rows to be deleted. 
This is an example: 
```
      MERGE INTO {catalog}.{silver_database}.purchase_promotion r
      USING tab_update u
      ON r.purchase_promotion_row_code = u.purchase_promotion_row_code  -- Match on promotion_row_code
      AND r.promotion_code = u.promotion_code
      -- If a row in the target (r) is not matched by any row in the source (u)
      WHEN NOT MATCHED BY SOURCE
      AND r.promotion_end_date > CAST('{date_from}' AS DATE)
      THEN
        UPDATE SET
          r.update_date = CAST('{update_date}' AS TIMESTAMP),
          r.to_be_deleted = 1
```
We understand only the delete action is supported at the moment. 
I contacted Ben from databricks and he suggested reaching out to you for this. I would love to connect with you to see if there is a way to include this logic in the merge conditions and actions improvement you developed. Do you have an email address I can contact you on?
Thanks in advance :)

**Comment by mi-volodin** [comment]
Hi, @sabrinas-tuidi 

Yes, I am afraid you are right. 

The reason I haven't included this is the management of update conditions. I didn't have the valid use case at hand, and also the `when match then update` clause didn't give me much of reusable code, because I won't expect that the whole row will be updated on delete.

Having your case, I think it won't be that difficult to extend the code if we assume that update action shall be stated explicitly in the same fashion the conditions are provided.

@benc-db what is the best branch to fork from for this extension?

**Comment by sabrinas-tuidi** [comment]
> Hi, @sabrinas-tuidi
> 
> Yes, I am afraid you are right.
> 
> The reason I haven't included this is the management of update conditions. I didn't have the valid use case at hand, and also the `when match then update` clause didn't give me much of reusable code, because I won't expect that the whole row will be updated on delete.
> 
> Having your case, I think it won't be that difficult to extend the code if we assume that update action shall be stated explicitly in the same fashion the conditions are provided.
> 
> **upd** Regarding the contact - I changed my privacy settings (🤷‍♂️ already signed by email my commits anyway) and also added linkedin link, feel free to text me.
> 
> @benc-db what is the best branch to fork from for this extension?

Hi @mi-volodin, thank you so much for your reply. I sent you a request on LinkedIn, it would be great to work with you on this, my colleagues and I are more than happy to discuss the best way to move forward. Let me know! :)
